docs: add mui tutorial and example (#1847)
pankod/refine: documentation/docs/ui-frameworks/mui/tutorial.md

Let's begin by setting up a new **refine** project.

## Setting up

There are two alternative methods to set up a **refine** application.

The recommended way is using the [superplate](https://github.com/pankod/superplate) tool. _superplate_'s _CLI wizard_ will let you create and customize your application in seconds.

Alternatively, you may use the _create-react-app_ tool to create an empty _React_ application and then add **refine** module via _npm_.

<Tabs
defaultValue="superplate"
values={[
{label: 'use superplate', value: 'superplate'},
{label: 'use create-react-app', value: 'create-react-app'}
]}>
<TabItem value="create-react-app">

First, run the _create-react-app_ tool to bootstrap an empty _React project_.

```
npx create-react-app tutorial --template typescript
```

:::note
_--template typescript_ flag will ensure that the Typescript language is selected.
:::
Navigate to the project folder and install **refine** with the following _npm_ command:

<Tabs
defaultValue="npm"
values={[
{label: 'use npm', value: 'npm'},
{label: 'use yarn', value: 'yarn'}
]}>
<TabItem value="npm">

```bash

> npm i @pankod/refine-core @pankod/refine-mui @pankod/refine-react-router-v6

Reviewer: ```suggestion
npm i @pankod/refine-core @pankod/refine-mui @pankod/refine-react-router-v6 notistack
```
I think, you can add `notistack` for notification provider.

Author: I think that it should export from @pankod/refine-mui. What do you think? @ozkalai @omeraplak

Reviewer 2: @biskuvit yes, you are right!

```

  </TabItem>
    <TabItem value="yarn">

```bash

> yarn add  @pankod/refine-core @pankod/refine-mui @pankod/refine-react-router-v6

Reviewer: ```suggestion
yarn add  @pankod/refine-core @pankod/refine-mui @pankod/refine-react-router-v6 notistack
```

Now, create a new folder named _"pages/posts"_ under _"/src"_. Under that folder, create a _"list.tsx"_ file with the following code:

```tsx title="pages/posts/list.tsx"
import {
    useDataGrid,
    DataGrid,
    GridColumns,
    TagField,
    DateField,
    List,
} from "@pankod/refine-mui";

import { IPost } from "interfaces";

const columns: GridColumns = [
    { field: "title", headerName: "Title", flex: 1, minWidth: 350 },
    {
        field: "status",
        headerName: "Status",
        minWidth: 150,
        renderCell: (params) => <TagField value={params.row.status} />,
    },
    {
        field: "createdAt",
        headerName: "CreatedAt",
        minWidth: 150,
        renderCell: (params) => (
            <DateField format="LLL" value={params.row.createdAt} />
        ),
    },
];

export const PostList: React.FC = () => {
    const { dataGridProps } = useDataGrid<IPost>({
        columns,
    });

    return (
        <List>
            <DataGrid {...dataGridProps} autoHeight />
        </List>
    );
};
```

<br/>

Let's break down the `<PostList/>` component to understand what's going on here:

> ✳️ `<DataGrid/>` is a native **Material UI** component. It renders records row by row as a table.

Reviewer 3: It could be better if we add a link to the DataGrid component, so users can easily reach this component on the Material UI page.

Author: Actually, I am doing that [here](https://github.com/pankod/refine/pull/1847/files/2eaa4bfa9882788ad7646126f917adbfb9391a12#diff-6e5041c14d6be0ddd6dd0a8e3d336b6d88111c05559352e8ce2f3d3c47391f7fR467)

`<DataGrid/>` expects a `columns` prop as a required.

:::note
**refine** uses [Material UI](https://mui.com/x/react-data-grid/#mit-version) components to render data.
You may refer to [Material UI Docs](https://mui.com/x/api/data-grid/data-grid/) for further information about the `<DataGrid/>` component.
:::

✳️ `useDataGrid<IPost>();` is passed to the `<DataGrid/>` component as `{...dataGridProps}`.

This is the point where the ✨real magic✨ happens!

> **refine** hook `useDataGrid()` fetches data from API and wraps them with various helper hooks required for the `<DataGrid/>` component. Data interaction functions like **sorting**, **filtering**, and **pagination** will be instantly available on the `<DataGrid/>` with this single line of code.

Reviewer 3: It could be better if we add a link to the useDataGrid hook, so users can easily reach this hook for looking deeply.

:::note
**refine** depends heavily on hooks and `useDataGrid()` is only one among many others.

> On [useTable() documentation](/core/hooks/useTable.md) you may find more information about the usage of this hook.

Reviewer: ```suggestion
On [useDataGrid() documentation](#) you may find more information about the usage of this hook.
```

```tsx title="src/pages/index.tsx"
export * from "./posts";
```

<br />

Similarly, put a file in the `/src/pages/posts` folder which accomplishes the same function. We will use the commented out code later as we add more capabilities to our app. Remember as you add functions, uncomment each appropriate line.

```tsx title="src/pages/posts/index.tsx"
export * from "./list";
// export * from "./edit";
// export * from "./create";
// export * from "./show";
```

<br />

Open your application in your browser. You will see **posts** are displayed correctly in a table structure and even the pagination works out-of-the box.

On the next step, we are going to add a category field to the table which involves handling data relationships.

<>

<div class="img-container">
    <div class="window">
        <div class="control red"></div>
        <div class="control orange"></div>
        <div class="control green"></div>
    </div>
    <img src="" alt="Resource only List component" />
</div>
<br/>
</>

<br/>

## Handling relationships

> Remember the records from `/posts` endpoint that had a category id field?

Reviewer 3: question mark

```ts title="https://api.fake-rest.refine.dev/posts/1"
...
  "category": {
    "id": 26
  }
...
```

<br />

To display category titles on our table, we first need to map category id's to their corresponding titles.
The category title data can be obtained from the `/categories` endpoint for each record.

```ts title="https://api.fake-rest.refine.dev/categories/26"
  {
    "id": 26,
    "title": "mock category title",
  }
```

<br />

> At this point, we need to join records from different resources. For this, we're goint to use the refine hook `useMany`.

Reviewer 3: ```suggestion
At this point, we need to join records from different resources. For this, we're going to use the refine hook `useMany`.
```

goint -> going

import { ICategory, IPost } from "interfaces";

export const PostShow: React.FC = () => {
    const { queryResult } = useShow<IPost>();

    const { data, isLoading } = queryResult;
    const record = data?.data;

    const { data: categoryData } = useOne<ICategory>({
        resource: "categories",
        id: record?.category.id || "",
        queryOptions: {
            enabled: !!record?.category.id,
        },
    });

    return (
        <Show isLoading={isLoading}>
            <Stack gap={1}>
                <Typography variant="body1" fontWeight="bold">
                    Title
                </Typography>
                <Typography variant="body2">{record?.title}</Typography>
                <Typography variant="body1" fontWeight="bold">
                    Status
                </Typography>
                <Typography variant="body2">{record?.status}</Typography>
                <Typography variant="body1" fontWeight="bold">
                    Category
                </Typography>
                <Typography variant="body2">
                    {categoryData?.data.title}
                </Typography>
            </Stack>
        </Show>
    );
};
```

> ✳️ `useShow()` is a **refine** hook used to fetch a single record data. The `queryResult` has the response and also `isLoading` state.

Reviewer 3: ```suggestion
✳️ `useShow()` is a **refine** hook used to fetch a single record of data. The `queryResult` has the response and also `isLoading` state.
```

[Refer to the `useShow` documentation for detailed usage information. &#8594](/core/hooks/show/useShow.md)

✳️ To retrieve the category title, again we need to make a call to `/categories` endpoint. This time we used `useOne()` hook to get a single record from another resource.

[Refer to the `useOne` documentation for detailed usage information. &#8594](/core/hooks/data/useOne.md)

:::caution attention

> `useShow()` is the preferred hook for fetching data from current resource. For query foreign resources you may use the low-level `useOne()` hook.

Reviewer 3: ```suggestion
`useShow()` is the preferred hook for fetching data from the current resource. To query foreign resources you may use the low-level `useOne()` hook.
```

:::

> Since we've got access to raw data returning from `useShow()`, there is no restriction how it's displayed on your components. If you prefer presenting your content with a nicer wrapper, **refine** provides you

Reviewer 3: ```suggestion
Since we've got access to raw data returning from `useShow()`, there is no restriction on how it's displayed on your components. If you prefer presenting your content with a nicer wrapper, **refine** provides you
```

            minWidth: 250,
            renderCell: (params) => {
                return (
                    <ShowButton
                        hideText
                        size="small"
                        variant="outlined"
                        recordItemId={params.row.id}
                    />
                );
            },
        },
        // highlight-end
    ];

    const { dataGridProps } = useDataGrid<IPost>({
        columns,
    });

    const {
        filterMode,
        filterModel,
        onFilterModelChange,
        ...restDataGridProps
    } = dataGridProps;

    return (
        <List>
            <DataGrid
                {...restDataGridProps}
                filterMode={filterMode}
                filterModel={filterModel}
                onFilterModelChange={onFilterModelChange}
                autoHeight
            />
        </List>
    );
};
```

<br />

<div class="img-container">
    <div class="window">
        <div class="control red"></div>
        <div class="control orange"></div>
        <div class="control green"></div>
    </div>
    <img src="" alt="Show record action" />
</div>
<br/>

## `Editing` a record

> Until this point, we were basically working with read operations such as fetching and displaying data from resources. From now on, we are going to start creating and updating records by using **refine**.

Reviewer 3: ```suggestion
Until this point, we were basically working with reading operations such as fetching and displaying data from resources. From now on, we are going to start creating and updating records by using **refine**.
```

                                              item.toString(),
                                      )?.title ?? "";
                            }}
                            isOptionEqualToValue={(option, value) =>
                                value === undefined ||
                                option.id.toString() === value.toString()
                            }
                            renderInput={(params) => (
                                <TextField
                                    {...params}
                                    label="items"
                                    margin="normal"
                                    variant="outlined"
                                    error={!!errors.item}
                                    helperText={errors.item && "item required"}
                                    required
                                />
                            )}
                        />
                    )}
                />
            </Box>
        </Edit>
    );
};
```
<br />

Let's see what's going on our `<PostEdit>` component in detail:

✳️ `useForm` is a refine hook for handling form data.

> On the example it returns `formProps` and `saveButtonProps`, where the former includes all necessary props to build the form and the latter has the ones for the save button.

Reviewer 3: ```suggestion
In the example, it returns `formProps` and `saveButtonProps`, where the former includes all necessary props to build the form and the latter has the ones for the save button.
```

:::caution Attention
In edit page, `useForm` hook initializes the form with current record values.

[Refer to the `useForm` documentation for detailed usage information . &#8594](/ui-frameworks/mui/hooks/form/useForm.md)

✳️ You can give form property to `<Box>` component and it will render the form.

✳️ `<TextField>` is Material UI components to build form inputs.

✳️ `<Autocomplete>` is a text input that helps you find what you're looking for by suggesting options. You may refer to the [Field Components Documentation](/ui-frameworks/mui/components/fields/autocomplete.md) to get the full information about `<Autocomplete>`, 

✳️ `<Edit>` is a wrapper **refine** component for `<form>`. It provides save, delete and refresh buttons that can be used for form actions.

✳️ Form data is set automatically with `register` coming out of the`useForm` hook, whenever children inputs `<TextField>`'s are edited.

> ✳️ Save button submits the form by executing the `useUpdate` method provided by the [`dataProvider`](/core/providers/data-provider.md). After a succesfull response, the application will be redirected to the listing page.

Reviewer 3: ```suggestion
✳️ Save button submits the form by executing the `useUpdate` method provided by the [`dataProvider`](/core/providers/data-provider.md). After a successful response, the application will be redirected to the listing page.
```

    LightTheme,
    CssBaseline,
    ThemeProvider,
} from "@pankod/refine-mui";
import dataProvider from "@pankod/refine-simple-rest";
import routerProvider from "@pankod/refine-react-router-v6";

import { authProvider } from "authProvider";
// highlight-next-line
import { PostEdit, PostList, PostShow } from "pages/posts";

const API_URL = "https://api.fake-rest.refine.dev";
const App: React.FC = () => {
    const notificationProvider = notificationProviderHandle();

    return (
        <ThemeProvider theme={LightTheme}>
            <CssBaseline />
            <Refine
                authProvider={authProvider}
                routerProvider={routerProvider}
                dataProvider={dataProvider(API_URL)}
                ReadyPage={ReadyPage}
                Layout={Layout}
                catchAll={<ErrorComponent />}
                resources={[
                    {
                        name: "posts",
                        list: PostList,
                        // highlight-next-line
                        edit: PostEdit,
                        show: PostShow,
                    },
                ]}
            />
        </ThemeProvider>
    );
};
```

> We are going to need an _edit_ button on each row to diplay the `<PostEdit>` component. **refine** doesn't automatically add one, so we have to update our `<PostList>` component to add a `<EditButton>` for each record:

Reviewer 3: ```suggestion
We are going to need an _edit_ button on each row to display the `<PostEdit>` component. **refine** doesn't automatically add one, so we have to update our `<PostList>` component to add a `<EditButton>` for each record:
```

} from "@pankod/refine-mui";
import dataProvider from "@pankod/refine-simple-rest";
import routerProvider from "@pankod/refine-react-router-v6";

import { authProvider } from "authProvider";
// highlight-next-line
import { PostCreate, PostEdit, PostList, PostShow } from "pages/posts";

const API_URL = "https://api.fake-rest.refine.dev";
const App: React.FC = () => {
    const notificationProvider = notificationProviderHandle();

    return (
        <ThemeProvider theme={LightTheme}>
            <CssBaseline />
            <Refine
                authProvider={authProvider}
                routerProvider={routerProvider}
                dataProvider={dataProvider(API_URL)}
                ReadyPage={ReadyPage}
                Layout={Layout}
                catchAll={<ErrorComponent />}
                resources={[
                    {
                        name: "posts",
                        list: PostList,
                        // highlight-next-line
                        create: PostCreate,
                        edit: PostEdit,
                        show: PostShow,
                    },
                ]}
            />
        </ThemeProvider>
    );
};
```

> And that's it! Try it on browser and see if you can create new posts from scratch.

Reviewer 3: ```suggestion
And that's it! Try it on the browser and see if you can create new posts from scratch.
```

<br />

<div class="img-container">
    <div class="window">
        <div class="control red"></div>
        <div class="control orange"></div>
        <div class="control green"></div>
    </div>
    <img src="" alt="Create record action" />
</div>

<br/>

## Deleting a record

Deleting a record can be done in two ways.

> First way is adding an delete button on each row since _refine_ doesn't automatically add one, so we have to update our `<PostList>` component to add a `<DeleteButton>` for each record:

Reviewer 3: ```suggestion
The first way is adding a delete button on each row since _refine_ doesn't automatically add one, so we have to update our `<PostList>` component to add a `<DeleteButton>` for each record:
```
